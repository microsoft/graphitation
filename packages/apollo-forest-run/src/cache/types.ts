import type { Cache, InMemoryCacheConfig, TypePolicies } from "@apollo/client";
import type { IndexedForest, IndexedTree } from "../forest/types";
import type {
  NodeKey,
  FieldName,
  OperationDescriptor,
  PossibleTypes,
  PossibleSelection,
  ArgumentValues,
  Directives,
  TypeName,
  OperationId,
} from "../descriptor/types";
import type { DocumentNode } from "graphql";
import type {
  Key,
  KeySpecifier,
  ObjectKey,
  SourceCompositeList,
  SourceObject,
} from "../values/types";
import type {
  FieldMergeFunction,
  FieldReadFunction,
} from "@apollo/client/cache/inmemory/policies";

export type DataTree = IndexedTree & {
  grown?: boolean;
};
export type ResultTree = DataTree & {
  danglingReferences?: Set<string>; // ApolloCompat
};

export type TransformedResult = {
  outputTree: ResultTree;
  dirtyNodes: Map<NodeKey, Set<FieldName>>;
};

export type ExtendedForest = IndexedForest & {
  // Derived trees resulted from applying read policies
  readResults: Map<OperationDescriptor, TransformedResult>;
  mutations: Set<OperationDescriptor>;

  // ApolloCompat:
  operationsWithDanglingRefs: Map<NodeKey, Set<OperationDescriptor>>;
};
export type DataForest = ExtendedForest & {
  layerTag: null;
  replay?: undefined;
};
export type OptimisticLayer = ExtendedForest & {
  layerTag: string;
  replay: <T>(cache: T) => any; // ApolloCompat
};

export type Store = {
  operations: Map<DocumentNode, Set<OperationDescriptor>>;
  dataForest: DataForest;
  optimisticLayers: OptimisticLayer[];
  // Derived trees resulted after applying optimistic layers and read policies
  optimisticReadResults: Map<OperationDescriptor, TransformedResult>;
  partialReadResults: Set<OperationDescriptor>; // FIXME: this should be per layer for proper cleanup
  watches: Map<OperationDescriptor, Array<Cache.WatchOptions>>;
  // Last access time of operation
  //   Used for LRU eviction
  atime: Map<OperationId, number>;
};

export type Transaction = {
  optimisticLayer: OptimisticLayer | null;
  affectedOperations: Set<OperationDescriptor> | null;
  watchesToNotify: Set<Cache.WatchOptions> | null;
  forceOptimistic: boolean | null;
  writes: {
    options: Cache.WriteOptions;
    tree: IndexedTree;
  }[];
};

export type CacheConfig = InMemoryCacheConfig & {
  maxOperationCount?: number;
  nonEvictableQueries?: Set<string>;
  apolloCompat_keepOrphanNodes?: boolean;
};

export type CacheEnv = {
  addTypename?: boolean; // ApolloCompat
  apolloCompat_keepOrphanNodes?: boolean;

  rootTypes?: {
    query: string;
    mutation: string;
    subscription: string;
  };

  possibleTypes?: PossibleTypes;

  now(): number;
  genId: () => number;

  // TODO: use keyFields instead of objectKey function
  //  (reason being - we can understand when some list contains items with "ids" even when the list is empty or currently contains nulls only)
  //  + possible to warn when some query doesn't include keyFields
  objectKey: (
    obj: object,
    fields?: PossibleSelection,
    operation?: OperationDescriptor,
  ) => ObjectKey | false | undefined;

  listItemKey?: (
    item: SourceObject | SourceCompositeList,
    index: number,
    // TODO:
    //   parentType: TypeName
    //   parentFieldName: string
  ) => string | number;

  // TODO: rename to fieldConfig
  keyArgs?: (
    typeName: string,
    fieldName: string,
    args?: ArgumentValues,
    directives?: Directives,
    source?: OperationDescriptor | undefined,
  ) => Key | KeySpecifier | undefined;

  // ApolloCompat:
  //   Apollo can track dirty entries in results of read operations even if some "key" fields are missing in selection
  //   by maintaining optimism dependency graph between main store and read results.
  //   ForestRun does indexing and diffing over actual "read results", so if the result selection doesn't contain
  //   key fields - it may be unable to properly identify an object in the result.
  //   To ensure compatibility with Apollo - we have to maintain this keyMap of literal objects to node "keys"
  //   for derived objects (e.g. created via reading selections)
  //   This is mostly needed for compatibility with Apollo tests - real world apps _shouldn't_ be affected
  keyMap?: WeakMap<SourceObject, string | false>;

  typePolicies: TypePolicies;
  dataIdFromObject?: any;

  mergePolicies: Map<TypeName, Map<FieldName, FieldMergeFunction>>;
  readPolicies: Map<TypeName, Map<FieldName, FieldReadFunction>>;
  nonEvictableQueries?: Set<string>;
  maxOperationCount?: number;
};
