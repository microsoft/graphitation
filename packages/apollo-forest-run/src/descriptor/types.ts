import {
  DocumentNode,
  FieldNode,
  FragmentDefinitionNode,
  FragmentSpreadNode,
  GraphQLFormattedError,
  InlineFragmentNode,
  OperationDefinitionNode,
  ValueNode,
} from "graphql";

export type TypeName = string;
export type FieldName = string;
export type FieldAlias = string;
export type DataKey = string;
export type FragmentName = string;
export type SelectedInOperation = true;
export type VariableName = string;
export type SelectedIn = FragmentName | SelectedInOperation;
export type DeferLabel = string;
export type VariableValues = { [name: VariableName]: unknown };
export type NodeKey = string;
export type OperationId = number;

export type ArgumentValues = Map<string, unknown>;
export type Directives = Map<string, { args: ArgumentValues }>;
export type KeySpecifier = readonly string[];
export type Key = string;

export type NormalizedFieldEntry =
  | string
  | {
      readonly name: FieldName;
      readonly args: ArgumentValues;
      readonly keyArgs?: Key | KeySpecifier;
    };

export type FragmentMap = Map<string, FragmentDefinitionNode>;

export type OperationEnv = {
  genId?: () => OperationId;

  // ApolloCompat:
  //   Have to keep it as a function for now, should convert to simple static list from type policies
  objectKey?: (
    obj: object,
    selection?: PossibleSelection,
    operation?: OperationDescriptor,
  ) => Key | false | undefined;

  keyArgs?: (
    typeName: string,
    fieldName: string,
    args?: ArgumentValues,
    directives?: Directives,
    source?: OperationDescriptor | undefined,
  ) => Key | KeySpecifier | undefined;
};

export type DocumentDescriptor = {
  document: DocumentNode;
  fragmentMap: FragmentMap;
  debugName: string;
  definition: OperationDefinitionNode;
};

export type ResultTreeDescriptor = {
  possibleSelections: PossibleSelections;
  fieldsWithArgs: FieldInfo[]; // aggregated from all levels
};

// Selection with additional information based on operation variables (resolved at runtime)
export type ResolvedSelection = PossibleSelection & {
  normalizedFields?: Map<FieldInfo, NormalizedFieldEntry>;
  skippedFields?: Set<FieldInfo>;
};

export type OperationDescriptor = {
  id: OperationId;
  env: OperationEnv;
  document: DocumentNode;
  fragmentMap: FragmentMap;
  debugName: string;
  definition: OperationDefinitionNode;
  possibleSelections: PossibleSelections;
  variables: VariableValues;
  variablesWithDefaults: VariableValues;
  keyVariables: VariableName[] | null;
  rootType: TypeName; // e.g. Query | Mutation, etc
  rootNodeKey: NodeKey; // e.g. ROOT_QUERY | ROOT_MUTATION, etc
  selections: Map<PossibleSelections, Map<TypeName | null, ResolvedSelection>>;
  cache: boolean;
};

export type FormattedError = GraphQLFormattedError;

export type FieldInfo = {
  name: FieldName;
  dataKey: FieldAlias | FieldName;
  args?: Map<string, ValueNode>; // keeping as ValueNode to simplify matching against variables later
  alias?: FieldAlias;
  selection?: PossibleSelections;
  selectedIn: SelectedIn[];

  // AST references MUST NOT be used outside indexing. There is no guarantee they will exist.
  // E.g. consider case when "Selection" structures are generated at build time.
  __refs: ASTReference[];
};

// Note: the same field node may have multiple references via different fragment spreads
export type ASTReference = {
  node: FieldNode;
  parentSpreads: (FragmentSpreadNode | InlineFragmentNode)[];
};

export type PossibleSelections = Map<TypeName | null, PossibleSelection>;
export type PossibleSelection = {
  depth: number;
  fields: FieldMap;
  fieldsWithSelections?: FieldName[];
  fieldQueue: FieldInfo[];
  fieldsToNormalize?: FieldInfo[];
  fieldsWithDirectives?: FieldInfo[];

  fragmentSpreads?: FragmentName[];
  experimentalAlias?: DataKey;
  experimentalAliasedFragments?: Map<DataKey, PossibleSelection>;
};

export type ResolvedSelections = Map<
  PossibleSelections,
  Map<TypeName | null, ResolvedSelection>
>;

export type PossibleTypes = { [abstractType: TypeName]: TypeName[] };

// e.g. { a: foo { bar }, { b: foo { baz } } -> same canonical name `foo` leads to multiple entries: `a { bar }` and `b { baz }`;
export type FieldMap = Map<FieldName, FieldInfo[]>;
export type FragmentSpreadMap = Map<FragmentName, SelectedIn[]>;
