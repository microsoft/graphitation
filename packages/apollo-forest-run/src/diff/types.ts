import {
  GraphValue,
  ObjectChunk,
  FieldName,
  CompositeListValue,
  ObjectValue,
  SourceObject,
  SourceCompositeList,
} from "../values/types";
import { FieldInfo, NormalizedFieldEntry } from "../descriptor/types";

export type DiffEnv = {
  allowMissingFields?: boolean;

  listItemKey?: (
    item: SourceObject | SourceCompositeList,
    index: number,
    // TODO:
    //   parentType: TypeName
    //   parentFieldName: string
  ) => string | number;
};

export type DiffContext = {
  env: DiffEnv;
  errors: DiffError[] | undefined;
};

export type ObjectDiffState = {
  difference?: ObjectDifference;
  errors?: DiffError[];
  allowMissingFields?: boolean;
};

export type DiffState = {
  difference?: ValueDifference;
  errors?: (DiffError | MissingModelError)[];
  allowMissingFields?: boolean;
};

export const enum DiffErrorKind {
  MissingModelValue,
  MissingModelFields,
  MissingBaseFields,
}

export type MissingModelError = {
  kind: DiffErrorKind.MissingModelValue;
};

export type MissingModelFieldsError = {
  kind: DiffErrorKind.MissingModelFields;
  chunk: ObjectChunk;
  missingFields: FieldInfo[];
};

export type MissingBaseFieldsError = {
  kind: DiffErrorKind.MissingBaseFields;
  chunk: ObjectChunk;
  missingFields: FieldInfo[];
};

export type DiffFieldError = MissingModelFieldsError | MissingBaseFieldsError;

export type DiffError =
  | MissingModelError
  | MissingModelFieldsError
  | MissingBaseFieldsError;

/**
 * Contains current diffing state.
 *
 * Since diff is executed against multiple object chunks, we may end up comparing the same field multiple times.
 * Information about diffing against previous chunks helps us skip unnecessary work and makes diffing more efficient.
 *
 * This state is supposed to be short-lived: usually within one event loop tick, while we are diffing a single
 * model object against all other base objects.
 *
 * State is an implementation detail of diffing algorithm and should not be relied upon by consumers: it can change at any time
 */
export const enum DifferenceKind {
  Replacement,
  Filler,
  ObjectDifference,
  CompositeListDifference,
  FieldEntryDifference,
}

export type ObjectDifference = {
  readonly kind: DifferenceKind.ObjectDifference;
  // readonly allFields: Iterable<FieldName>;

  fieldQueue: Set<FieldName>;
  fieldState: Map<FieldName, FieldEntryDifference | FieldEntryDifference[]>;
  dirtyFields?: Set<FieldName>;
  errors?: DiffError[];
};

export type FieldEntryDifference = {
  readonly kind: DifferenceKind.FieldEntryDifference;
  fieldEntry: NormalizedFieldEntry;
  state: ValueDifference;
};

export type CompositeListDifference = {
  readonly kind: DifferenceKind.CompositeListDifference;
  itemQueue: Set<number>;
  itemState: Map<number, ValueDifference>;
  dirtyItems?: Set<number>;
  layout?: CompositeListLayoutDifference;
  deletedKeys?: string[];
  errors?: DiffError[];
};

type AddedValue = CompositeListValue | ObjectValue;
export type CompositeListLayoutDifference = (number | null | AddedValue)[];

export type Replacement = {
  readonly kind: DifferenceKind.Replacement;
  oldValue: GraphValue;
  newValue: GraphValue;
};

export type Filler = {
  readonly kind: DifferenceKind.Filler;
  readonly newValue: GraphValue;
};

export type ValueDifference =
  | ObjectDifference
  | CompositeListDifference
  | Replacement
  | Filler;

export type NodeDifferenceMap = Map<string, ObjectDifference>;
