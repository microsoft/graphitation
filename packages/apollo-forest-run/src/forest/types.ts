import {
  FieldInfo,
  OperationDescriptor,
  OperationId,
  PossibleSelection,
  PossibleSelections,
  TypeName,
} from "../descriptor/types";
import {
  CompositeListChunk,
  DataMap,
  MissingFieldsMap,
  NodeKey,
  NodeMap,
  ObjectChunk,
  ObjectDraft,
  ObjectKey,
  ObjectValue,
  OperationResult,
  ParentLocator,
  SourceCompositeList,
  SourceObject,
  TypeMap,
} from "../values/types";
import { TelemetryEvent } from "../telemetry/types";
import { Logger } from "../jsutils/logger";
import { UpdateTreeStats } from "../telemetry/updateStats/types";
import { UpdateLogger } from "../telemetry/updateStats/updateLogger";

export type IndexedTree = {
  operation: OperationDescriptor;
  result: OperationResult;
  rootNodeKey: NodeKey;
  nodes: NodeMap;
  typeMap: TypeMap;

  // Inverse map from data to chunks
  //   This is necessary for efficient re-indexing of existing data and chunk recycling
  dataMap: DataMap;

  // Previous state of this tree.
  //   Useful for recycling chunks from multiple tree states, (e.g. with merge policies when final tree
  //   should recycle chunks from both - incoming and existing trees).
  prev: IndexedTree | null;

  // Error states
  incompleteChunks: Set<ObjectChunk | CompositeListChunk>;

  // ApolloCompat
  danglingReferences?: Set<NodeKey>;
};

export type IndexedForest = {
  trees: Map<OperationId, IndexedTree>;
  extraRootIds: Map<NodeKey, TypeName>;
  operationsByNodes: Map<NodeKey, Set<OperationId>>; // May contain false positives
  operationsWithErrors: Set<OperationDescriptor>; // May contain false positives
  deletedNodes: Set<NodeKey>;
};

export type Source = Readonly<SourceObject | SourceCompositeList>;
export type Draft = SourceObject | SourceCompositeList;

export type UpdateForestStats = (UpdateTreeStats | null)[];

// Changed chunks map only contains chunks with immediate changes (i.e. "Replacement", "Filler" + list layout changes).
//   Does not contain parent chunks which were affected only because some nested chunk has changed.
//   Note: For now dirty list items are not reported, as it is tricky to report together with list layout shifts (and we don't need it anywhere yet).
//         In the future we may need to report layout shifts and "Replacement", "Fillter" changes separately.
export type ChangedChunksMap = Map<ObjectChunk, FieldInfo[]> &
  Map<CompositeListChunk, null>;

export type UpdateTreeContext = {
  operation: OperationDescriptor;
  drafts: Map<Source, Draft>;
  missingFields: MissingFieldsMap;
  changes: ChangedChunksMap;
  changedNodes: Set<NodeKey>;
  affectedNodes: Set<NodeKey>;
  completeObject: CompleteObjectFn;
  findParent: ParentLocator;
  logger: ForestEnv["logger"];
  statsLogger?: UpdateLogger;
};

export type UpdateTreeResult = {
  updatedTree: IndexedTree;
  changes: ChangedChunksMap;
  changedNodes: Set<NodeKey>; // Directly changed nodes (subset of NodeDifferenceMap keys)
  affectedNodes: Set<NodeKey>; // Parent nodes updated due to a change in a nested node (if node is both - directly updated and affected by another node update, it will be in both: changedNodes and affectedNodes)
  stats?: UpdateTreeStats;
};

export type UpdateObjectResult = {
  draft: SourceObject;
  missingFields?: MissingFieldsMap;
};

export type CompleteObjectFn = (
  value: ObjectValue,
  selection: PossibleSelections,
  operation: OperationDescriptor,
) => ObjectDraft;

export type ForestEnv = {
  objectKey: (
    obj: SourceObject,
    fields?: PossibleSelection,
    operation?: OperationDescriptor,
  ) => ObjectKey | false | undefined;

  logger?: Logger;
  notify?: (event: TelemetryEvent) => void;

  // ApolloCompat:
  //   Apollo can track dirty entries in results of read operations even if some "key" fields are missing in selection
  //   by maintaining optimism dependency graph between main store and read results.
  //   ForestRun does indexing and diffing over actual "read results", so if the result selection doesn't contain
  //   key fields - it may be unable to properly identify an object in the result.
  //   To ensure compatibility with Apollo - we have to maintain this keyMap of literal objects to node "keys"
  //   for derived objects (e.g. created via reading selections)
  //   This is mostly needed for compatibility with Apollo tests - real world apps _shouldn't_ be affected
  keyMap?: WeakMap<SourceObject, string | false>;
  apolloCompat_keepOrphanNodes?: boolean;

  // Telemetry feature flags
  logUpdateStats?: boolean;
  logStaleOperations?: boolean;
};
