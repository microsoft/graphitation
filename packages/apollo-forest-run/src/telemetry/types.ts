import { UpdateTreeStats } from "./updateStats/types";

// IMPORTANT!
// 1. Events MUST NOT include any user data or any other information that can be used to identify a user.
//    This is necessary to comply with GDPR and other privacy regulations.
//    Events could be sent to the server and may be stored in logs or other storage.
// 2. Events MUST be serializable with JSON.stringify.
type OperationDebugName = string;
type FirstMissingFieldMessage = string;

// Event occurs when incoming write causes missing fields in watched operations.
// This happens due to a mismatching selection in the incoming operation and the affected operation.
//
// There are several situations when this can happen:
// - Incoming operation adds a new list item, but affected operation has some field in the selection
//     which is not present in the incoming operation, so it is impossible to complete the new array item.
//
// - Incoming operation switches field value of abstract type from one specific type to anoter (i.e. "__typename" changes),
//     but selection for another type of the affected operation contains fields in selection
//     which are not present in the incoming operation, so it is impossible to complete the object.
//
// - Incoming operation changes relationship between objects (i.e. changes value of some field
//   from object with id "A" to object with id "B"), but affected operation has fields in selection
//     which are not present in the incoming operation,
//
export type UnexpectedRefetch = {
  kind: "UNEXPECTED_REFETCH";
  causedBy: OperationDebugName[];
  affected: Array<[OperationDebugName, FirstMissingFieldMessage]>;
};

// Event occurs when read policy throws an error
export type ReadPolicyError = {
  kind: "READ_POLICY_ERROR";
  op: OperationDebugName;
  type: string;
  field: string;
};

// Event occurs when merge policy throws an error
export type MergePolicyError = {
  kind: "MERGE_POLICY_ERROR";
  op: OperationDebugName;
  type: string;
  field: string;
};

// Event occurs when merge policy throws an error
export type UpdateStats = {
  kind: "UPDATE_STATS";
  causedBy: OperationDebugName;
  updateStats: UpdateTreeStats[];
};

export type TelemetryEvent =
  | UnexpectedRefetch
  | UpdateStats
  | ReadPolicyError
  | MergePolicyError;
