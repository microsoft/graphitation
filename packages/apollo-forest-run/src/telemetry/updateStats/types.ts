import { TypeName } from "../../descriptor/types";

export type UpdateTreeStats = {
  operationName: string;
  updates: ChunkUpdateStats[];
} & CopyStats;

export type CopyStats = {
  arraysCopied: number;
  arrayItemsCopied: number;
  objectFieldsCopied: number;
  objectsCopied: number;
};

/**
 * @property nodeType - The type of the mutated node.
 * @property depth - The depth of the node in the tree.
 * @property fieldsMutated - Number of object fields changed in this mutation.
 * @property itemsMutated - Number of array items changed in this mutation.
 * @property updateStats - Detailed stats for this mutation, including fields/items mutated and copy stats.
 * @property parentUpdateStats - (Optional) Copy stats for changes propagated up to the root, if applicable.
 */
export type ChunkUpdateStats = {
  nodeType: TypeName;
  depth: number;
  updateStats: {
    fieldsMutated: number;
    itemsMutated: number;
  } & CopyStats;
  updateAscendantStats?: CopyStats;
};
