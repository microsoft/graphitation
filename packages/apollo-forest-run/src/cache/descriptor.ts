import type { DocumentNode } from "graphql";
import type { CacheEnv, Store } from "./types";
import type {
  DocumentDescriptor,
  OperationDescriptor,
  ResultTreeDescriptor,
  VariableName,
} from "../descriptor/types";
import { equal } from "@wry/equality";
import { describeDocument } from "../descriptor/document";
import { applyDefaultValues, describeOperation } from "../descriptor/operation";
import { addTypenameToDocument } from "../descriptor/addTypenameToDocument";
import { assert } from "../jsutils/assert";
import { describeResultTree } from "../descriptor/possibleSelection";
import { Cache } from "@apollo/client";
import { VariableValues } from "../descriptor/types";

export const ROOT_TYPES = Object.freeze(["Query", "Mutation", "Subscription"]);
export const ROOT_NODES = Object.freeze([
  "ROOT_QUERY",
  "ROOT_MUTATION",
  "ROOT_SUBSCRIPTION",
]);

const typenameDocumentCache = new WeakMap<DocumentNode, DocumentNode>();
const reverseDocumentCache = new WeakMap<DocumentNode, DocumentNode>();
const resultTreeDescriptors = new WeakMap<DocumentNode, ResultTreeDescriptor>();
const diffDescriptors = new WeakMap<Cache.DiffOptions, OperationDescriptor>();

export function resolveOperationDescriptor(
  env: CacheEnv,
  { operations, dataForest }: Store,
  doc: DocumentNode,
  variables?: { [key: string]: unknown },
  rootNodeKey?: string,
  keyVars: VariableName[] | null = null,
): OperationDescriptor & { rootNodeKey: string | false } {
  const document = env.addTypename ? transformDocument(doc) : doc;
  const documentDescriptor = describeDocument(document);

  let variants = operations.get(document);
  if (!variants) {
    variants = new Set<OperationDescriptor>();
    operations.set(document, variants);
  }

  let rootTypeName;
  if (isFragmentDocument(documentDescriptor)) {
    const fragment = getFragmentNode(documentDescriptor);
    rootTypeName = fragment.typeCondition.name.value;
  }

  const variablesWithDefaultValues = applyDefaultValues(
    variables ?? {},
    documentDescriptor.definition.variableDefinitions,
  );

  let match: OperationDescriptor | undefined;
  for (const variant of variants) {
    if (
      variablesAreEqual(
        variablesWithDefaultValues,
        variant.variablesWithDefaults,
        keyVars,
      ) &&
      (typeof rootNodeKey === "undefined" ||
        variant.rootNodeKey === rootNodeKey)
    ) {
      match = variant;
      break;
    }
  }
  if (!match) {
    let resultTreeDescriptor = resultTreeDescriptors.get(document);
    if (!resultTreeDescriptor) {
      resultTreeDescriptor = describeResultTree(
        documentDescriptor,
        env.possibleTypes,
      );
      resultTreeDescriptors.set(document, resultTreeDescriptor);
    }
    match = describeOperation(
      env,
      documentDescriptor,
      resultTreeDescriptor,
      variables ?? {},
      variablesWithDefaultValues,
      rootTypeName,
      rootNodeKey,
    );
    variants.add(match);
  }
  if (
    typeof rootNodeKey !== "undefined" &&
    !ROOT_NODES.includes(rootNodeKey) &&
    ROOT_TYPES.includes(match.rootType)
  ) {
    match.rootType = dataForest.extraRootIds.get(rootNodeKey) ?? match.rootType;
  }
  return match;
}

export function transformDocument(document: DocumentNode): DocumentNode {
  let result = typenameDocumentCache.get(document);
  if (!result) {
    result = addTypenameToDocument(document);
    typenameDocumentCache.set(document, result);
    // If someone calls transformDocument and then mistakenly passes the
    // result back into an API that also calls transformDocument, make sure
    // we don't keep creating new query documents.
    typenameDocumentCache.set(result, result);
    reverseDocumentCache.set(result, document);
  }
  return result;
}

export function getOriginalDocument(
  maybeTransformed: DocumentNode,
): DocumentNode {
  return reverseDocumentCache.get(maybeTransformed) ?? maybeTransformed;
}

export function isFragmentDocument(descriptor: DocumentDescriptor) {
  const operationDefinition = descriptor.definition;
  const selections = operationDefinition.selectionSet.selections;

  return selections.length === 1 && selections[0].kind === "FragmentSpread";
}

export function getFragmentNode(descriptor: DocumentDescriptor) {
  const fragmentSpreadNode = descriptor.definition.selectionSet.selections[0];
  assert(fragmentSpreadNode.kind === "FragmentSpread");

  const fragment = descriptor.fragmentMap.get(fragmentSpreadNode.name.value);
  if (!fragment) {
    throw new Error(`No fragment named ${fragmentSpreadNode.name.value}.`);
  }
  return fragment;
}

export function getDiffDescriptor(
  env: CacheEnv,
  store: Store,
  options: Cache.WatchOptions | Cache.DiffOptions,
): OperationDescriptor {
  // Diff / watch options could be used interchangeably multiple times with the same watch | diff object
  let operationDescriptor = diffDescriptors.get(options);
  if (!operationDescriptor) {
    operationDescriptor = resolveOperationDescriptor(
      env,
      store,
      options.query,
      options.variables as any,
      options.rootId ?? options.id, // ApolloCompat: both variants are used in different scenarios ü§∑‚Äç
    );
    diffDescriptors.set(options, operationDescriptor);
  }
  return resolveKeyDescriptor(env, store, operationDescriptor);
}

/**
 * Returned result will be the same as `operation` when `@cache(keyVars: [...])` is not set on document,
 * otherwise - the first found descriptor with matching keyVars
 */
export function resolveKeyDescriptor(
  env: CacheEnv,
  store: Store,
  operation: OperationDescriptor,
): OperationDescriptor {
  return operation.keyVariables
    ? resolveOperationDescriptor(
        env,
        store,
        operation.document,
        operation.variables,
        operation.rootNodeKey,
        operation.keyVariables,
      )
    : operation;
}

function variablesAreEqual(
  a: VariableValues,
  b: VariableValues,
  keyVars: string[] | null = null,
): boolean {
  if (!keyVars) {
    return equal(a, b);
  }
  for (const name of keyVars) {
    if (!equal(a[name], b[name])) {
      return false;
    }
  }
  return true;
}
