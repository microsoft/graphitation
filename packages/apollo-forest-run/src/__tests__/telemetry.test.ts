import { gql } from "../__tests__/helpers/descriptor";
import { ForestRun } from "../ForestRun";
import { UpdateStats, UnexpectedRefetch } from "../telemetry/types";

describe("Update Stats Telemetry", () => {
  const query = gql`
    query MyQuery {
      user {
        id
        name
        email
        friends {
          id
          name
          profile {
            id
            bio
          }
        }
      }
    }
  `;

  const initialData = {
    user: {
      __typename: "User",
      id: "1",
      name: "Alice",
      email: "alice@example.com",
      friends: [
        {
          __typename: "Friend",
          id: "2",
          name: "Bob",
          profile: {
            __typename: "Profile",
            id: "BobProfile",
            bio: "Default bio for Bob",
          },
        },
        {
          __typename: "Friend",
          id: "3",
          name: "Charlie",
          profile: {
            __typename: "Profile",
            id: "CharlieProfile",
            bio: "Default bio for Charlie",
          },
        },
      ],
    },
  };

  const updateMutation = gql`
    mutation UpdateFriend {
      updateFriend {
        __typename
        ... on User {
          id
          name
          email
        }
        ... on Profile {
          id
          bio
        }
        ... on Friend {
          id
          name
        }
      }
    }
  `;

  const updatedData = {
    updateFriend: [
      {
        __typename: "User",
        id: "1",
        name: "Updated Alice",
        email: "updated@example.com",
      },
      {
        __typename: "Profile",
        id: "BobProfile",
        bio: "Updated bio for Bob",
      },
      {
        __typename: "Friend",
        id: "2",
        name: "Updated Bob",
      },
    ],
  };
  it("should report update stats telemetry", () => {
    const telemetryEvents: UpdateStats[] = [];
    const cache = new ForestRun({
      logUpdateStats: true,
      notify: (event) => {
        if (event.kind === "UPDATE_STATS") {
          telemetryEvents.push(event);
        }
      },
    });

    const diffs: any[] = [];

    cache.watch({
      query,
      optimistic: true,
      callback: (diff) => {
        diffs.push(diff);
      },
    });

    cache.write({ query, result: initialData });
    cache.write({
      query: updateMutation,
      result: updatedData,
    });

    expect(telemetryEvents.length).toEqual(1);
    expect(telemetryEvents).toMatchSnapshot();
  });

  it("should not report update stats telemetry when disabled", () => {
    const telemetryEvents: UpdateStats[] = [];
    const cache = new ForestRun({
      logUpdateStats: false,
      notify: (event) => {
        if (event.kind === "UPDATE_STATS") {
          telemetryEvents.push(event);
        }
      },
    });

    cache.watch({
      query,
      optimistic: true,
      callback: () => {},
    });

    cache.write({ query, result: initialData });
    cache.write({
      query: updateMutation,
      result: updatedData,
    });

    expect(telemetryEvents.length).toEqual(0);
  });
});

describe("Unexpected refetch telemetry", () => {
  const query = gql`
    query GetUserWithDetails {
      user {
        id
        name
        friends {
          id
          name
        }
      }
    }
  `;

  const initialData = {
    user: {
      __typename: "User",
      id: "1",
      name: "Alice",
      friends: [
        {
          __typename: "Friend",
          id: "2",
          name: "Bob",
        },
      ],
    },
  };

  const fragmentWithMissingFields = gql`
    fragment PartialUser on User {
      id
      name
      friends {
        name
      }
    }
  `;

  const fragmentData = {
    __typename: "User",
    friends: [
      {
        __typename: "Friend",
        name: "Updated Bob",
      },
    ],
  };

  it("should report unexpected refetch telemetry", () => {
    const telemetryEvents: UnexpectedRefetch[] = [];

    const cache = new ForestRun({
      notify: (event) => {
        if (event.kind === "UNEXPECTED_REFETCH") {
          telemetryEvents.push(event);
        }
      },
      logStaleOperations: true,
    });

    cache.watch({
      query: query,
      optimistic: true,
      callback: () => {},
    });

    cache.write({
      query: query,
      result: initialData,
    });

    cache.writeFragment({
      id: cache.identify({ __typename: "User", id: "1" })!,
      fragment: fragmentWithMissingFields,
      data: fragmentData,
    });

    expect(telemetryEvents).toMatchSnapshot();
  });

  it("should not report unexpected refetch telemetry when disabled", () => {
    const telemetryEvents: UnexpectedRefetch[] = [];

    const cache = new ForestRun({
      notify: (event) => {
        if (event.kind === "UNEXPECTED_REFETCH") {
          telemetryEvents.push(event);
        }
      },
    });

    cache.watch({
      query: query,
      optimistic: true,
      callback: () => {},
    });

    cache.write({
      query: query,
      result: initialData,
    });

    cache.writeFragment({
      id: cache.identify({ __typename: "User", id: "1" })!,
      fragment: fragmentWithMissingFields,
      data: fragmentData,
    });

    expect(telemetryEvents.length).toEqual(0);
  });
});
