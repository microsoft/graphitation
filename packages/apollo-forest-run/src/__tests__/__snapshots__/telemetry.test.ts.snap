// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unexpected refetch telemetry should report unexpected refetch telemetry 1`] = `
[
  {
    "affected": [
      [
        "query GetUserWithDetails",
        "user.friends.0.id",
      ],
    ],
    "causedBy": [
      "query {
  ...PartialUser
}",
    ],
    "kind": "UNEXPECTED_REFETCH",
  },
]
`;

exports[`Update Stats Telemetry should report update stats telemetry 1`] = `
[
  {
    "causedBy": "mutation UpdateFriend",
    "kind": "UPDATE_STATS",
    "updateStats": [
      {
        "arrayItemsCopied": 2,
        "arraysCopied": 1,
        "objectFieldsCopied": 13,
        "objectsCopied": 4,
        "operationName": "query MyQuery",
        "updates": [
          {
            "depth": 3,
            "nodeType": "Profile",
            "updateAscendantStats": {
              "arrayItemsCopied": 2,
              "arraysCopied": 1,
              "objectFieldsCopied": 10,
              "objectsCopied": 3,
            },
            "updateStats": {
              "arrayItemsCopied": 0,
              "arraysCopied": 0,
              "fieldsMutated": 1,
              "itemsMutated": 0,
              "objectFieldsCopied": 3,
              "objectsCopied": 1,
            },
          },
          {
            "depth": 2,
            "nodeType": "Friend",
            "updateAscendantStats": {
              "arrayItemsCopied": 2,
              "arraysCopied": 1,
              "objectFieldsCopied": 6,
              "objectsCopied": 2,
            },
            "updateStats": {
              "arrayItemsCopied": 0,
              "arraysCopied": 0,
              "fieldsMutated": 1,
              "itemsMutated": 0,
              "objectFieldsCopied": 4,
              "objectsCopied": 1,
            },
          },
          {
            "depth": 1,
            "nodeType": "User",
            "updateAscendantStats": {
              "arrayItemsCopied": 0,
              "arraysCopied": 0,
              "objectFieldsCopied": 1,
              "objectsCopied": 1,
            },
            "updateStats": {
              "arrayItemsCopied": 0,
              "arraysCopied": 0,
              "fieldsMutated": 2,
              "itemsMutated": 0,
              "objectFieldsCopied": 5,
              "objectsCopied": 1,
            },
          },
        ],
      },
    ],
    "watchersCount": 1,
  },
]
`;
